extensions/guacamole-auth-radius/src/main/java/org/apache/guacamole/auth/radius/AuthenticationProviderService.java

```diff
@@ -0,0 +1,223 @@
+            }
+
+            // If we get back an AccessReject packet, login is denied.
+            else if (radPack instanceof AccessReject) {
+                logger.debug("Login has been rejected by RADIUS server.");
+                throw new GuacamoleInvalidCredentialsException("Permission denied.", CredentialsInfo.USERNAME_PASSWORD);
+            }
+
+            // If we receive an AccessChallenge package, the server needs more information -
+            // We create a new form/field with the challenge message.
+            else if (radPack instanceof AccessChallenge) {
+
+                // Try to get the state attribute - if it's not there, we have a problem
+                RadiusAttribute stateAttr = radPack.findAttribute(Attr_State.TYPE);
+                if (stateAttr == null) {
+                    logger.error("Something went wrong, state attribute not present.");
+                    logger.debug("State Attribute turned up null, which shouldn't happen in AccessChallenge.");
+                    throw new GuacamoleInvalidCredentialsException("Authentication error.", CredentialsInfo.USERNAME_PASSWORD);
+                }
+
+                // We need to get the reply message so we know what to ask the user
+                RadiusAttribute replyAttr = radPack.findAttribute(Attr_ReplyMessage.TYPE);
+                if (replyAttr == null) {
+                    logger.error("No reply message received from the server.");
+                    logger.debug("Expecting a Attr_ReplyMessage attribute on this packet, and did not get one.");
+                    throw new GuacamoleInvalidCredentialsException("Authentication error.", CredentialsInfo.USERNAME_PASSWORD);
+                }
+
+                // We have the required attributes - convert to strings and then generate the additional login box/field
+                String replyMsg = replyAttr.toString();
+                String radiusState = new String(stateAttr.getValue().getBytes());
+                Field radiusResponseField = new RadiusChallengeResponseField(replyMsg);
+                Field radiusStateField = new RadiusStateField(radiusState);
+                CredentialsInfo expectedCredentials = new CredentialsInfo(Arrays.asList(radiusResponseField,radiusStateField));
+                throw new GuacamoleInsufficientCredentialsException("LOGIN.INFO_RADIUS_ADDL_REQUIRED", expectedCredentials);
+
+            }
+
+            // If we receive AccessAccept, authentication has succeeded
+            else if (radPack instanceof AccessAccept) {
+
+                try {
+
+                    // Return AuthenticatedUser if bind succeeds
+                    AuthenticatedUser authenticatedUser = authenticatedUserProvider.get();
+                    authenticatedUser.init(credentials);
+                    return authenticatedUser;
+                }
+                finally {
+                    radiusService.disconnect();
```

Comment:
I'm not 100% certain that JRadius cleans up after itself, and since I'm doing a lot of the RADIUS calls manually to do the challenge/response authentication, I wanted to be sure that the RADIUS connection was cleaned up when the authentication completed.  I assume by "only being cleaned up here" you mean that, if it's being cleaned up, here, there are potentially some other places where it might need to be cleaned up, so I'm missing a few calls to disconnect()?